csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependencyManager.cs

```diff
@@ -736,11 +730,21 @@ private void DownloadMissingPackages(List<FileInfo> allFiles, ISet<string> dllPa
                     return;
                 }
 
-                dotnet.RestoreProjectToDirectory(tempDir.DirInfo.FullName, missingPackageDirectory.DirInfo.FullName, forceDotnetRefAssemblyFetching: false, out var _, pathToNugetConfig: nugetConfig);
-                // TODO: the restore might fail, we could retry with a prerelease (*-* instead of *) version of the package.
+                success = dotnet.RestoreProjectToDirectory(tempDir.DirInfo.FullName, missingPackageDirectory.DirInfo.FullName, forceDotnetRefAssemblyFetching: false, out var _, out var outputLines, pathToNugetConfig: nugetConfig);
                 if (!success)
                 {
-                    progressMonitor.FailedToRestoreNugetPackage(package);
+                    if (outputLines?.Any(s => s.Contains("NU1301")) == true)
+                    {
+                        // Restore could not be completed because the listed source is unavailable. Try without the nuget.config:
+                        success = dotnet.RestoreProjectToDirectory(tempDir.DirInfo.FullName, missingPackageDirectory.DirInfo.FullName, forceDotnetRefAssemblyFetching: false, out var _, out var _, pathToNugetConfig: null, force: true);
```

Comment:
This is just a suggestions and nothing that should block merging: Since we are now adding a second out parameter (which contains all output lines) and since we have several ignores of the list of assets files (which is calculated eagerly in the call to `dotnet.RestoreProjectToDirectory`) it might make sense to introduce a new class to carry the result of a "restore" call.
Maybe something like (`RestoreResult` should be the return type of `RestoreProjectToDictionary` and `RestoreSolutionToDictionary`.
```csharp
        public class RestoreResult
        {
            private readonly IList<string> output;
            public bool Success { get; }

            public RestoreResult(bool success, IList<string>? output)
            {
                Success = success;
                this.output = output ?? Array.Empty<string>();
            }

            private static IEnumerable<string> GetFirstGroupOnMatch(Regex regex, IEnumerable<string> lines) =>
                lines
                    .Select(line => regex.Match(line))
                    .Where(match => match.Success)
                    .Select(match => match.Groups[1].Value);

            public IEnumerable<string> GetAssetsFilePaths() => GetFirstGroupOnMatch(AssetsFileRegex(), output);

            public IEnumerable<string> GetRestoredProjects() => GetFirstGroupOnMatch(RestoredProjectRegex(), output);

            public bool HasNugetError() => output.Any(s => s.Contains("NU1301"));
        }
```

Reply:
Good idea, I'll do it in a followup PR.